processing/src/test/java/org/apache/druid/query/UnnestScanQueryRunnerTest.java

```diff
@@ -0,0 +1,530 @@
+  public static Iterable<Object[]> constructorFeeder()
+  {
+
+    return QueryRunnerTestHelper.cartesian(
+        QueryRunnerTestHelper.makeUnnestQueryRunners(
+            FACTORY,
+            QueryRunnerTestHelper.PLACEMENTISH_DIMENSION,
+            QueryRunnerTestHelper.PLACEMENTISH_DIMENSION_UNNEST,
+            null
+        ),
+        ImmutableList.of(false, true)
+    );
+  }
+
+  private Druids.ScanQueryBuilder newTestUnnestQuery()
+  {
+    return Druids.newScanQueryBuilder()
+                 .dataSource(QueryRunnerTestHelper.UNNEST_DATA_SOURCE)
+                 .columns(Collections.emptyList())
+                 .eternityInterval()
+                 .limit(3)
+                 .legacy(legacy);
+  }
+
+  private Druids.ScanQueryBuilder newTestUnnestQueryWithAllowSet()
+  {
+    List<String> allowList = Arrays.asList("a", "b", "c");
+    LinkedHashSet allowSet = new LinkedHashSet(allowList);
+    return Druids.newScanQueryBuilder()
+                 .dataSource(UnnestDataSource.create(
+                     new TableDataSource(QueryRunnerTestHelper.DATA_SOURCE),
+                     QueryRunnerTestHelper.PLACEMENTISH_DIMENSION,
+                     QueryRunnerTestHelper.PLACEMENTISH_DIMENSION_UNNEST,
+                     allowSet
+                 ))
+                 .columns(Collections.emptyList())
+                 .eternityInterval()
+                 .limit(3)
+                 .legacy(legacy);
+  }
+
+  @Test
+  public void testUnnestRunner()
+  {
+    ScanQuery query = newTestUnnestQuery()
+        .intervals(I_0112_0114)
+        .columns(QueryRunnerTestHelper.PLACEMENTISH_DIMENSION_UNNEST)
+        .limit(3)
+        .build();
+
+    Iterable<ScanResultValue> results = runner.run(QueryPlus.wrap(query)).toList();
+    String[] columnNames;
+    if (legacy) {
+      columnNames = new String[]{
+          getTimestampName() + ":TIME",
+          QueryRunnerTestHelper.PLACEMENTISH_DIMENSION_UNNEST
+      };
+    } else {
+      columnNames = new String[]{
+          QueryRunnerTestHelper.PLACEMENTISH_DIMENSION_UNNEST
+      };
+    }
+    String[] values;
+    if (legacy) {
+      values = new String[]{
+          "2011-01-12T00:00:00.000Z\ta",
+          "2011-01-12T00:00:00.000Z\tpreferred",
+          "2011-01-12T00:00:00.000Z\tb"
+      };
+    } else {
+      values = new String[]{
+          "a",
+          "preferred",
+          "b"
+      };
+    }
+
+    final List<List<Map<String, Object>>> events = toEvents(columnNames, values);
+    List<ScanResultValue> expectedResults = toExpected(
+        events,
+        legacy
+        ? Lists.newArrayList(getTimestampName(), QueryRunnerTestHelper.PLACEMENTISH_DIMENSION_UNNEST)
+        : Collections.singletonList(QueryRunnerTestHelper.PLACEMENTISH_DIMENSION_UNNEST),
+        0,
+        3
+    );
+    ScanQueryRunnerTest.verify(expectedResults, results);
+  }
+
+  @Test
+  public void testUnnestRunnerVirtualColumns()
```

Comment:
i think would also be interesting to have a test on a composite array column using `array` function to create an array from multiple columns (in addition to `mv_to_array` test)